File: go/2024/mon_4/30/500. Keyboard Row.go
package main

import (
	"fmt"
	"strings"
)

func findWords(words []string) []string {
	// 定义一个目标 []string
	var desStrList []string
	// 构建不同的 set (Go语言没有set，这里使用map实现相同的结果)
	var zeroSet = make(map[int32]bool)
	for _, char := range "qwertyuiop" {
		zeroSet[char] = true
	}

	var oneSet = make(map[int32]bool)
	for _, char := range "asdfghjkl" {
		oneSet[char] = true
	}

	var twoSet = make(map[int32]bool)
	for _, char := range "zxcvbnm" {
		twoSet[char] = true
	}
	// 遍历 word，判断是否在多个集合里面
	for i, word := range words {
		// 对每一个单词进行判定
		word = strings.ToLower(word)
		var wordSet = make(map[int32]int)
		var canAdd bool = true
		for _, char := range word {
			if _, ok := zeroSet[char]; ok {
				wordSet[0]++
			} else if _, ok := oneSet[char]; ok {
				wordSet[1]++
			} else {
				wordSet[2]++
			}
			if len(wordSet) > 1 {
				canAdd = false
				break
			}
		}
		if canAdd {
			desStrList = append(desStrList, words[i])
		}
	}
	return desStrList
}

func main() {
	fmt.Println(findWords([]string{"Hello", "Alaska", "Dad", "Peace"}))
}

File: python/2024/mon_4/30/500. Keyboard Row.py
import collections
from typing import List, Dict, Set


class Solution:
    def findWords(self, words: List[str]) -> List[str]:
        """
        T(n): O(n)
        S(n): O(1)
        """
        # 构建 row_dict
        letter_row_dic: Dict[str, int] = collections.defaultdict(int)
        for letter in "qwertyuiop":
            letter_row_dic[letter] = 0
        for letter in "asdfghjkl":
            letter_row_dic[letter] = 1
        for letter in "zxcvbnm":
            letter_row_dic[letter] = 2
        des_str_list: List[str] = []
        for word in words:
            lowercase_word = word.lower()
            diff_set: Set[int] = set()
            # 判断当前的 word 组成的 letter 来自哪些行
            for letter in lowercase_word:
                diff_set.add(letter_row_dic[letter])

            if len(diff_set) == 1:
                des_str_list.append(word)

        return des_str_list


if __name__ == "__main__":
    s = Solution()
    print(s.findWords(["Hello", "Alaska", "Dad", "Peace"]))
